### PR TITLE
Fix Unsafe Object Deseralization in Microsoft's App Center SDK

### PR DESCRIPTION
**Pull request by shadyabarada**
Things to consider before you submit the PR:

* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [x] Did you check UI tests on the sample app? They are not executed on CI.
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

The issue is basically that the usage of the NSKeyUnarchiever and NSKeyArchiever with "requireSecureCoding" set to No, allows the manipulation on the data being deserialised, in other words, you can be expecting to deserialise an object of type ArrayList but the manipulation gives you a Dictionary. This does not throw any errors, because requireSecureCoding is set to No. The solution is to enable secure coding for a given set of classes and all these classes should conform to NSSecureCoding. Thus no outsider data can be used to breach this.

## Related PRs or issues

[AB#95662](https://dev.azure.com/msmobilecenter/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/95662)

## Misc

Add what's missing, notes on what you tested, additional thoughts or questions.

### CODE REVIEW

**Comment by DmitriyKirakosyan** [comment]
@shadyabarada Fix formatting by running script `./clang-format-changed-files.sh`.

**Review by DmitriyKirakosyan — CHANGES_REQUESTED** [review]

File: AppCenter/AppCenter/Internals/Util/MSACUtility.m [hunk 1]
```diff
@@ -113,4 +118,8 @@ + (void)addMigrationClasses:(NSDictionary<NSString *, id> *)data {
   }
   [targetClasses addEntriesFromDictionary:data];
 }
+
++ (void)addAllowedClasses:( NSArray *)allowedClassesArray {
+    allowedClasses = allowedClassesArray;
```
[the remark this answers is not in the document]
Comment (shadyabarada) [reply AppCenter/AppCenter/Internals/Util/MSACUtility.m]: done

File: AppCenter/AppCenterTests/MSACAbstractLogTests.m [hunk 2]
```diff
@@ -34,6 +50,10 @@ - (void)setUp {
   self.sut.distributionGroupId = @"FAKE-GROUP-ID";
   self.sut.userId = @"FAKE-USER-ID";
   self.sut.device = OCMPartialMock([MSACDevice new]);
+  
+  NSArray *allowedClassesArray = @[[MSACAbstractLog class], [NSDate class], [MSACDevice class], [MSACDeviceHistoryInfo class], [MSACSessionHistoryInfo class], [MSACUserIdHistoryInfo class], [MSACAppExtension class], [MSACCommonSchemaLog class], [MSACCSData class], [MSACCSExtensions class], [MSACDeviceExtension class], [MSACLocExtension class], [MSACMetadataExtension class], [MSACNetExtension class], [MSACOSExtension class], [MSACProtocolExtension class], [MSACSDKExtension class], [MSACUserExtension class], [MSACStartServiceLog class], [MSACBooleanTypedProperty class], [MSACDateTimeTypedProperty class], [MSACDoubleTypedProperty class], [MSACLongTypedProperty class], [MSACStringTypedProperty class], [MSACTypedProperty class], [MSACHistoryInfo class], [MSACLogWithProperties class], [MSACWrapperSdk class], [NSUUID class], [NSDictionary class], [NSArray class], [NSNull class]];
```
[the remark this answers is not in the document]
Comment (shadyabarada) [reply AppCenter/AppCenterTests/MSACAbstractLogTests.m]: done

File: AppCenter/AppCenter/Internals/Util/MSACUtility.m [hunk 3]
```diff
@@ -54,13 +59,11 @@ + (NSObject *)unarchiveKeyedData:(NSData *)data {
   NSObject *unarchivedData;
   NSException *exception;
   @try {
-    if (@available(iOS 11.0, macOS 10.13, watchOS 4.0, *)) {
+    if (@available(macos 10.13, ios 11.0, watchos 4.0, tvOS 11.0, *)) {
```
[the remark this answers is not in the document]
Comment (shadyabarada) [reply AppCenter/AppCenter/Internals/Util/MSACUtility.m]: done

**Comment by shadyabarada** [comment]
@DmitriyKirakosyan the script for formatting return an error "clang-format: command not found
Failed to format file: AppCenter/AppCenter/Internals/Context/UserId/MSACUserIdHistoryInfo.m
"
for almost every file

File: AppCenterDistribute/AppCenterDistribute/Internals/MSACDistributeArchiverUtil.h [hunk 4]
```diff
@@ -0,0 +1,14 @@
+//
```
Comment (DmitriyKirakosyan) [inline AppCenterDistribute/AppCenterDistribute/Internals/MSACDistributeArchiverUtil.h:1]: Let's use annotations as in other files for the consistency.
```
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
```

File: AppCenterCrashes/AppCenterCrashes.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -1720,6 +1741,13 @@
 /* End PBXProject section */
 
 /* Begin PBXReferenceProxy section */
+		AFB5AB72296461FC002EBC38 /* CrashReporter.framework */ = {
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes.xcodeproj/project.pbxproj:1744]: Why was this added?

File: AppCenterCrashes/AppCenterCrashes.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -444,6 +450,13 @@
 			remoteGlobalIDString = C9A91F5D230BFDB70068070D;
 			remoteInfo = "AppCenterCrashes macOS Framework";
 		};
+		AFB5AB71296461FC002EBC38 /* PBXContainerItemProxy */ = {
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes.xcodeproj/project.pbxproj:453]: Why was this added?

File: AppCenter/AppCenterTests/MSACCommonSchemaLogTests.m [hunk 7]
```diff
@@ -41,6 +42,8 @@ - (void)setUp {
   } mutableCopy];
   [self.csLogDummyValues addEntriesFromDictionary:abstractDummies];
   self.commonSchemaLog = [self csLogWithDummyValues:self.csLogDummyValues];
+  // Set Allowed Classes
```
Comment (DmitriyKirakosyan) [inline AppCenter/AppCenterTests/MSACCommonSchemaLogTests.m:45]: empty line before comments

File: AppCenter/AppCenter/MSACAppCenter.m [hunk 8]
```diff
@@ -299,6 +300,8 @@ - (instancetype)init {
     @"MSWrapperSdk" : MSACWrapperSdk.self,
     @"MSAbstractLog" : MSACAbstractLog.self,
   }];
+  // Set Allowed Classes
```
Comment (DmitriyKirakosyan) [inline AppCenter/AppCenter/MSACAppCenter.m:303]: add empty line

**Review by DmitriyKirakosyan — CHANGES_REQUESTED** [review]

File: AppCenterAnalytics/AppCenterAnalytics.xcodeproj/project.pbxproj [hunk 9]
```diff
@@ -375,48 +378,6 @@
 			remoteGlobalIDString = C9A91F2B230BFB8A0068070D;
 			remoteInfo = "AppCenterAnalytics tvOS Framework";
 		};
-		D55735C6254AAE6100B64478 /* PBXContainerItemProxy */ = {
```
[the remark this answers is not in the document]
Comment (shadyabarada) [reply AppCenterAnalytics/AppCenterAnalytics.xcodeproj/project.pbxproj:378]: seems this was by mistake, will revert

File: AppCenterCrashes/AppCenterCrashes.xcodeproj/project.pbxproj (hunk 5 above)
Comment (shadyabarada) [reply AppCenterCrashes/AppCenterCrashes.xcodeproj/project.pbxproj:1744]: hmm seems like this was a local config, will revert

File: AppCenterDistribute/AppCenterDistribute/Internals/MSACDistributeArchiverUtil.m [hunk 10]
```diff
@@ -0,0 +1,27 @@
+//
```
Comment (DmitriyKirakosyan) [inline AppCenterDistribute/AppCenterDistribute/Internals/MSACDistributeArchiverUtil.m:1]: Fix annotations, same as in .h file.

File: AppCenterDistribute/AppCenterDistribute/MSACDistribute.m [hunk 11]
```diff
@@ -116,6 +117,8 @@ - (instancetype)init {
       @"MSDistributionStartSessionLog" : MSACDistributionStartSessionLog.self
     }];
```
Comment (DmitriyKirakosyan) [inline AppCenterDistribute/AppCenterDistribute/MSACDistribute.m:119]: Remove empty line

File: CHANGELOG.md [hunk 12]
```diff
@@ -4,6 +4,7 @@
 
 ### App Center
 
+* **[Fix]** Fix Unsafe Object Deserialization.
```
Comment (DmitriyKirakosyan) [inline CHANGELOG.md:7]: Please provide a link to the vulnerability which you are fixing

File: AppCenterCrashes/AppCenterCrashesTests/MSACBinaryTests.m [hunk 13]
```diff
@@ -12,6 +13,12 @@ @implementation MSACBinaryTests
 
 #pragma mark - Tests
 
+- (void)setUp {
+  [super setUp];
+
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashesTests/MSACBinaryTests.m:18]: remove empty line

File: AppCenterCrashes/AppCenterCrashesTests/MSACAppleErrorLogTests.m [hunk 14]
```diff
@@ -22,6 +23,8 @@ - (void)setUp {
   [super setUp];
 
   self.sut = [self appleErrorLog];
+
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashesTests/MSACAppleErrorLogTests.m:26]: remove empty line

File: AppCenterCrashes/AppCenterCrashes/Internals/Util/MSACCrashesArchiverUtil.m [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    [NSNull class],
+    [MSACLogWithProperties class],
+    [MSACCommonSchemaLog class],
+    [NSMutableData class],
+    [MSACExceptionModel class],
+    [NSString class],
+    [NSNumber class],
+    [NSMutableArray class]
+  ];
+
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes/Internals/Util/MSACCrashesArchiverUtil.m:50]: remove empty line

File: AppCenterCrashes/AppCenterCrashes.xcodeproj/project.pbxproj [hunk 16]
```diff
@@ -1832,10 +1846,10 @@
 			remoteRef = C24C9D022472B3C30072C35D /* PBXContainerItemProxy */;
 			sourceTree = BUILT_PRODUCTS_DIR;
 		};
-		C24C9D072472B3C30072C35D /* Tests tvOS.xctest */ = {
+		C24C9D072472B3C30072C35D /* CrashReporter tvOS Tests.xctest */ = {
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes.xcodeproj/project.pbxproj:1849]: I wonder why Xcode adds these changes..

File: AppCenterCrashes/AppCenterCrashes.xcodeproj/project.pbxproj [hunk 17]
```diff
@@ -1295,14 +1305,15 @@
 			isa = PBXGroup;
 			children = (
 				C24C9CED2472B3C30072C35D /* CrashReporter.framework */,
+				AFB5AB72296461FC002EBC38 /* CrashReporter.framework */,
```
Comment (DmitriyKirakosyan) [inline AppCenterCrashes/AppCenterCrashes.xcodeproj/project.pbxproj:1308]: this line doesn't look related to your changes.

File: AppCenterAnalytics/AppCenterAnalytics/MSACAnalytics.m [hunk 18]
```diff
@@ -74,6 +75,8 @@ - (instancetype)init {
       @"MSStartSessionLog" : MSACStartSessionLog.self
     }];
```
Comment (DmitriyKirakosyan) [inline AppCenterAnalytics/AppCenterAnalytics/MSACAnalytics.m:77]: remove empty line

File: AppCenterAnalytics/AppCenterAnalytics/Internals/Util/MSACAnalyticsArchiverUtil.m [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+    [MSACStringTypedProperty class],
+    [MSACTypedProperty class],
+    [MSACStartSessionLog class],
+    [NSDictionary class],
+    [MSACStartSessionLog class],
+    [NSString class],
+    [NSNumber class],
+    [NSMutableArray class]
+  ];
+
```
Comment (DmitriyKirakosyan) [inline AppCenterAnalytics/AppCenterAnalytics/Internals/Util/MSACAnalyticsArchiverUtil.m:53]: remove empty line

File: AppCenterAnalytics/AppCenterAnalytics/Internals/Util/MSACAnalyticsArchiverUtil.m [hunk 20]
```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#import <Foundation/Foundation.h>
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (DmitriyKirakosyan) [inline AppCenterAnalytics/AppCenterAnalytics/Internals/Util/MSACAnalyticsArchiverUtil.m:6]: We don't need two Copyrights😅

File: AppCenterAnalytics/AppCenterAnalytics.xcodeproj/project.pbxproj [hunk 21]
```diff
@@ -368,55 +371,55 @@
 			remoteGlobalIDString = C9A91F36230BFB970068070D;
 			remoteInfo = "AppCenterAnalytics macOS Framework";
 		};
+        D55735C6254AAE6100B64478 /* PBXContainerItemProxy */ = {
```
Comment (DmitriyKirakosyan) [inline AppCenterAnalytics/AppCenterAnalytics.xcodeproj/project.pbxproj:374]: Formatting

File: AppCenter/AppCenter/Internals/Util/MSACArchiverUtil.m [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+    [MSACWrapperSdk class],
+    [NSUUID class],
+    [NSDictionary class],
+    [NSArray class],
+    [NSNull class],
+    [NSString class],
+    [NSNumber class],
+    [NSMutableArray class]
+  ];
+
```
Comment (DmitriyKirakosyan) [inline AppCenter/AppCenter/Internals/Util/MSACArchiverUtil.m:74]: remove empty line

File: AppCenter/AppCenter/Internals/Util/MSACArchiverUtil.m [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+#import "MSACUserIdContext.h"
+#import "MSACUserIdHistoryInfo.h"
+#import "MSACUtility.h"
+#import "MSACWrapperSdk.h"
+#import <Foundation/Foundation.h>
+
+@implementation MSACArchiverUtil
+
++ (void)addAllowedAppCenterModuleClasses {
+
```
Comment (DmitriyKirakosyan) [inline AppCenter/AppCenter/Internals/Util/MSACArchiverUtil.m:36]: remove empty line

**Review by DmitriyKirakosyan — CHANGES_REQUESTED** [review]

File: AppCenterCrashes/AppCenterCrashes.xcodeproj/project.pbxproj (hunk 16 above)
Comment (shadyabarada) [reply AppCenterCrashes/AppCenterCrashes.xcodeproj/project.pbxproj:1849]: yeah this is weird

**Review by DmitriyKirakosyan — APPROVED** [review]
